### ISSUE TITLE
Container Encapsulation for ATLauncher 

### ISSUE THREAD

**Issue by GeoDerp** [ATLauncher/ATLauncher#769]
### The Problem

Greetings. Currently Everyone (including myself) is concerned about the latest virus coming out. Wouldn't it be nice to Encapsulate the whole application adding security to the host machine.

Im wondering if if its viable if we can create some documentation/modification for running ATLauncher via Docker (with Linux,Windows (WSL2) and Mac having different documentation )

### The Solution

This is very ruff draft idea. Im happy to work on it more in the future. But I just wanted to create something to get the idea across.
```docker
FROM redhat/ubi8 # or the nvidia cuda's version of ubi8, not sure about the legality however
RUN dnf install java-1.8.0-openjdk-devel curl git -y
RUN git clone https://github.com/ATLauncher/ATLauncher.git
WORKDIR /ATLauncher
RUN ./gradlew build #this currently fails the test
ENTRYPOINT jarfile=$(ls *.jar) && java -Xmx8g -Xms512m -jar $jarfile

# Example for windows via WSL2 (using ex. from https://stackoverflow.com/a/75392952) . Im assuming Mac and Linux to be alot easier
# $dir being the directory you want to copy instance data from

# docker build -t ATLauncher .
# docker run -v $dir:/ATLuncher/instances -v /run/desktop/mnt/host/wslg/.X11-unix:/tmp/.X11-unix -v /run/desktop/mnt/host/wslg:/mnt/wslg  -e DISPLAY=:0  -e WAYLAND_DISPLAY=wayland-0 -e XDG_RUNTIME_DIR=/mnt/wslg/runtime-dir -e PULSE_SERVER=/mnt/wslg/PulseServer ATLauncher
```

### Have you checked this issue doesn't already exist?

- [X] I am sure this issue doesn't exist already

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Doomsdayrs** [comment]
This is redundant for Linux as flatpak exists which provides containerization https://flathub.org/apps/com.atlauncher.ATLauncher

**Comment by RyanTheAllmighty** [comment]
There will likely be some effort involved in this, I know there's currently discussions going on by the Fractureiser Mitigation Team to look into and think about how launchers can effectively sandbox the Minecraft instance.

I imagine there will be something coming out of that, so until there it's hard to say for now how it will be done or if it can be done, but yeah I'll keep updates in here.

**Comment by GeoDerp** [comment]
This may be too complex (ill maybe see if I can get the ATLuncher running as a container this weekend), but I'm wondering. Instead of containerizing the whole launcher. Having a separate Image per minecraft instance. ATLuncher would then facilitate crafting the minecraft instance, crafting an Image and running the minecraft as a docker container. 

My idea of the procedure would go as follows:
- download and run ATLuncher 
- create a minecraft instance (select version, mods, etc). you can also tell minecraft to give the container a folder/volume that will store: world data, resource packs and Shaders. (ie. persistent storage)
- on play, ATLuncher will craft a Docker Image, if there is a docker image with the same hash (ie. existing image has been made with the same minecraft version and mod combination ) Docker will use the existing image on machine, otherwise docker will pull down the resources required and craft a new one.
- ATLuncher will then run the container using said Image, inserting variables like java memory limits and volumes. on minecraft close, the container will stop and be auto deleted.

Im more then happy to help start this process if this is something people are interested in.

**Comment by Doomsdayrs** [comment]
This requires docker to either be bundled (bloating app size) or to be preinstalled (for only one app??).

What is a better approach is to utilize native sandboxing tools. I know MacOS has it, and flatpak is already sandboxed. Windows on the other hand, I do not think has such features.

**Comment by GeoDerp** [comment]
> This requires docker to either be bundled (bloating app size) or to be preinstalled (for only one app??).
> 
> What is a better approach is to utilize native sandboxing tools. I know MacOS has it, and flatpak is already sandboxed. Windows on the other hand, I do not think has such features.

Yeah your right, You need to have WSL2 and Docker installed to run this on windows (Linux and Mac just Docker). I'm Kinda trying to find a approach that works on all operating Systems.  (I think this will probally be something that will be external from ATLuncher)
I'm putting more focus on Windows as I know there is already the Flatpak version for linux (thanks for mentioning this before). I'm not sure how the Flatpat works, but to my understanding a FLatpak opens up alot of holes by default to get it to run (like the home directory) I could be very wrong with this information tho. 
I'm trying to work out if there is a way to have as minimal holes opened as possible and have it work on every operating system or as many as possible.

**Comment by Doomsdayrs** [comment]
> I'm not sure how the Flatpat works, but to my understanding a FLatpak opens up alot of holes by default to get it to run (like the home directory) I could be very wrong with this information tho

The information is wrong in this case. For ATLauncher on FlatPak, the user download dir is read only (as of recently, previously write) and has access to send activity to discord. These are the only views out of the sandbox ATLauncher has by default.

![image](https://github.com/ATLauncher/ATLauncher/assets/38189170/f2065809-93d1-41e4-82f8-aa788af24af2)

(bulkGames etc is just a custom directory setup for ATLauncher)

---

Will be interesting to see what you create.

**Comment by GeoDerp** [comment]
> > I'm not sure how the Flatpat works, but to my understanding a FLatpak opens up alot of holes by default to get it to run (like the home directory) I could be very wrong with this information tho
> 
> The information is wrong in this case. For ATLauncher on Flatpak, the user download dir is read only (as of recently, previously write) and has access to send activity to discord. These are the only views out of the sandbox ATLauncher has by default.
> 
> ![image](https://user-images.githubusercontent.com/38189170/244842845-f2065809-93d1-41e4-82f8-aa788af24af2.png)
> 
> (bulkGames etc is just a custom directory setup for ATLauncher)
> 
> Will be interesting to see what you create.

Yeah thats sick! but yeah, We need to find a good Window and Mac alternative

**Comment by Doomsdayrs** [comment]
I do not think Windows has one for apps that are not UWPs.

For MacOS, there seems to be a way to sandbox apps natively. https://paolozaino.wordpress.com/2015/08/04/how-to-run-your-applications-in-a-mac-os-x-sandbox-to-enhance-security/

**Comment by Doomsdayrs** [comment]
If you want to implement per-instance sandboxing, [firejail](https://github.com/netblue30/firejail) can be used.

**Comment by RyanTheAllmighty** [comment]
I'd hold off doing anything until the Fractureiser Mitigation Team have their working group look into this more and come to some consensus or suggestions for launchers yo take.

**Comment by GeoDerp** [comment]
Sounds good, tempuarily however I have got a docker container "working" (lets call it alpha) for windows:
  
  Dockerfile 
  ```docker
    FROM redhat/ubi8 
    # or the nvidia cuda's version of ubi8, im not sure about the legality of this however
    
    #install Java 1.8, git 
    RUN dnf install java-1.8.0-openjdk-devel git -y

    # clone ATLuncher and build 
    RUN git clone https://github.com/ATLauncher/ATLauncher.git
    WORKDIR /ATLauncher
    # build and bypass test (as x11 wont work while docker is building )
    RUN ./gradlew build -x test
    
    #move built .jar to empty dir
    WORKDIR /ATLauncher/dist
    RUN mkdir -p ../run && jarfile=$(ls *.jar) && cp $jarfile ../run/ATLauncher.jar
    
    #Container run, runs ATLuncher.jar
    ENTRYPOINT java -Xmx$MEMORYLIMIT -Xms512m -jar /ATLauncher/run/ATLauncher.jar
```
**Commands to run:**
*(make sure you have Docker installed on Windows): https://www.docker.com/*
- make a new folder, in the folder past the script above and save the name as `Dockerfile`, then run terminal in this folder
- set $dir as directory you want to copy instance data to be stored from (ex. `mkdir instances` then replace $dir bellow with `./instances`)
- build image with `docker build -t atlauncher .`

* (using ex. from https://stackoverflow.com/a/75392952) *
- run image as container with command `docker run -e MEMORYLIMIT=8g -v $dir:/ATLauncher/run/instances -v /run/desktop/mnt/host/wslg/.X11-unix:/tmp/.X11-unix -v /run/desktop/mnt/host/wslg:/mnt/wslg  -e DISPLAY=:0  -e WAYLAND_DISPLAY=wayland-0 -e XDG_RUNTIME_DIR=/mnt/wslg/runtime-dir -e PULSE_SERVER=/mnt/wslg/PulseServer atlauncher `

**Current Issues:** 
Currently I am having issues with logging into the Microsoft account. Usually you open a browser, login, and the code gets sent to the ATLuncher application. Because we are running in a container and network wont be 127.0.0.1. The code dosent seem to get sent. I'm wondering if its possible to have a text box to place the code directly in. or, if you or @Doomsdayrs /Flatpak devs have found another solution to this issue.

Thanks for all the amazing work.
